### DISCUSSION TITLE
Can I --vimgrep with only 1 match per line?

### DISCUSSION THREAD

**Discussion by hberge** [BurntSushi/ripgrep#2282] [Q&A] [answered]
My use case is I am trying to use ripgrep (v13) as vim's `grepprg` to populate the quickfix list and then make substitutions with the quickfix list. E.g. in vim:
```
:set grepprg=rg\ --vimgrep\ $*
:grep pattern1
:cdo :s/pattern2/prefix_pattern2/g
```
This works as expected, but my issue is when a line has multiple occurences of pattern1 the `:cdo` command is executed multiple times resulting in multiple prefixes. The vim-internal `:vimgrep` command has a flag `g` for grep patterns, which toggles between adding single or multiple entries per line to the quickfix list. Is there any way to get the same function with the --vimgrep option in ripgrep?

Since I am probably not permitted to upgrade ripgrep on the remote where I am working I am also interested in work-arounds. The ones I can think of on my own are, so far:

1. In many cases I could be using `:cfdo :%s/patt/repl/gc` as this will execute once per file even if the quickfix list has multiple matches.
2. It would be possible to filter the results to one match per line. Maybe there are some vimscript plugins that can help?

Since this is my first post, thank you so much for this amazing tool! It's really spot-on for what's necessary in a developer's toolkit!

**Comment by hberge** [comment, marked as answer]
Wow, only seconds after posting I found the `--no-heading` option. So with ripgrep v13 I can get the behavior I want by setting grepprg in vim as follows:
```
:set grepprg=rg\ -C\ 0\ -n\ --no-heading\ $*
```

**Reply by BurntSushi** [reply]
Right. `--no-heading` gives you the standard grep format. The *point* of `--vimgrep` is that it gives you every single match, even if that means printing lines multiple times. I don't use vim's built-in search functionality, so I don't know _why_ Vim does things that way, but that's what it normally expects. So indeed, if you don't want to print the same line for every match, then don't use `--vimgrep`, because that's the only way to get that behavior in the first place. Every other aspect of `--vimgrep` can be controlled with other options, as you discovered.

Thank for you posting your solution. :-)
